### PR TITLE
TLS Support

### PR DESCRIPTION
**Pull request by lchudinov**
TLS Support for HTTP Server using gskit library.

### CODE REVIEW

**Review by daveyc — APPROVED** [review]
LGTM

File: c/tls.c [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+}
+
+int tlsDestroy(TlsEnvironment *env) {
+  int rc = 0;
+  rc = gsk_environment_close(env->envHandle);
+  safeFree((char*)env, sizeof(*env));
+  return rc;
+}
+
+static int waitTimeMs = 200;
```
Comment (ifakhrutdinov) [inline c/tls.c:51]: Perhaps making this variable `const` and using an all-caps name would better show the intention.

File: c/tls.c [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    while (1) {
+      PollItem item = {0};
+      item.fd = fd;
+      item.events = POLLERDNORM;
+      int returnCode = 0;
+      int reasonCode = 0;
+      int status = fdPoll(&item, 0, 1, waitTimeMs, &returnCode, &reasonCode);
+      if (status == -1) {
+        continue;
+      }
```
Comment (ifakhrutdinov) [inline c/tls.c:65]: Can we guarantee we'll leave this loop? What if things get broken and `status` is always `-1`?

**Review by ifakhrutdinov — COMMENTED** [review]
@lchudinov please see my comments. As for the tls code, it'd be great if someone who's worked with it reviewed it, perhaps Jared?

File: c/bpxskt.c [hunk 3]
```diff
@@ -1595,7 +1626,16 @@ int socketClose(Socket *socket, int *returnCode, int *reasonCode){
     returnValue = rs_ssl_releaseConnection(socket->sslHandle);
     return returnValue;
   }
-#endif
+#endif // USE_RS_SSL
+#ifdef USE_ZOWE_TLS
+  if (!socket->isServer && socket->tlsSocket) {
```
[the remark this answers is not in the document]
Comment (lchudinov) [reply c/bpxskt.c:1631]: Actually, `socket->tlsSocket` may be not null only on peer socket. I could've put `assert()` here.

File: h/tls.h [hunk 4]
```diff
@@ -0,0 +1,62 @@
+/*
```
[the remark this answers is not in the document]
Comment (lchudinov) [reply h/tls.h:1]: yes, I added #error directive in case of Metal C.

File: c/tls.c [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+#include <stdlib.h>
+#include <errno.h>
+#include "alloc.h"
+#include "bpxnet.h"
+#include "fdpoll.h"
+#include "tls.h"
+
+int tlsInit(TlsEnvironment **outEnv, TlsSettings *settings) {
+  int rc = 0;
+  TlsEnvironment *env = (TlsEnvironment *)safeMalloc(sizeof(*env), "Tls Environment");
```
[the remark this answers is not in the document]
Comment (lchudinov) [reply c/tls.c:19]: Added a check for NULL.

File: c/tls.c (hunk 1 above)
Comment (lchudinov) [reply c/tls.c:54]: Done

File: c/bpxskt.c [hunk 6]
```diff
@@ -998,6 +1001,20 @@ int socketRead(Socket *socket, char *buffer, int desiredBytes,
     }
   }
 #endif
+#ifdef USE_ZOWE_TLS
```
[the remark this answers is not in the document]
Comment (lchudinov) [reply c/bpxskt.c:1004]: They are mutually exclusive.

File: c/httpserver.c [hunk 7]
```diff
@@ -1570,6 +1573,60 @@ HttpServer *makeSecureHttpServer(STCBase *base, int port,
...
+  SessionTokenKey sessionTokenKey = {0};
+  if (initSessionTokenKey(&sessionTokenKey) != 0) {
+    return NULL;
+  }
+
+  Socket *listenerSocket = tcpServer2(addr,port,tlsFlags,returnCode,reasonCode);
+  if (listenerSocket == NULL){
+    return NULL;
+  }
+  listenerSocket->tlsEnvironment = tlsEnv;
```
[the remark this answers is not in the document]
Comment (lchudinov) [reply c/httpserver.c]: No problem. I refactored the code.

File: c/tls.c (hunk 2 above)
Comment (lchudinov) [reply c/tls.c]: I improved the loop.

**Review by ifakhrutdinov — APPROVED** [review]

**Comment by ifakhrutdinov** [comment]
This feature will need to be thoroughly tested since people may start relying on this.

**Pull request merged**